cou/apps/subordinate.py

```diff
@@ -0,0 +1,80 @@
+#  Copyright 2023 Canonical Limited
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+"""Subordinate application class."""
+import logging
+
+from cou.apps.app import SUBORDINATES, AppFactory, OpenStackApplication
+from cou.exceptions import ApplicationError
+from cou.steps import UpgradeStep
+from cou.utils.openstack import OpenStackRelease
+
+logger = logging.getLogger(__name__)
+
+
+@AppFactory.register_application(SUBORDINATES)
+class OpenStackSubordinateApplication(OpenStackApplication):
+    """Subordinate application class."""
+
+    def __post_init__(self) -> None:
+        """Initialize the Application dataclass."""
+        self.channel = self.status.charm_channel
+        self.charm_origin = self.status.charm.split(":")[0]
+        self.os_origin = self._get_os_origin()
+
+    @property
+    def current_os_release(self) -> OpenStackRelease:
+        """Assume latest version since principal already upgraded the packages."""
```

Comment:
This docstring is not explanatory. We could change it to something like:
```suggestion
        """Infer the OS release from subordinate charm's channel.
        
        We cannot determine the OS release base on workload packages because the principal charm has already upgraded the packages.
        :return: OpenStackRelease object.
        :rtype: OpenStackRelease
        """
```

Also added the `:return` variable descirption in docstring

Reply:
done